File: mobile/android/USDTgWallet/app/src/main/java/com/usdtgverse/wallet/data/BonusManager.kt
/*
==============================================
 File:        BonusManager.kt
 Author:      USDTG GROUP TECHNOLOGY LLC
 Created:     2025-10-13
 Last Update: 2025-10-13
 Version:     1.0

 Description:
   USDTgVerse Android Bonus Manager
   
   Manages purchase bonus system:
   - 10,000 USDTg → +10 USDTg bonus
   - VIP tier system (Bronze to Diamond)
   - Automatic bonus distribution

 License:
   MIT License
==============================================
*/

package com.usdtgverse.wallet.data

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

class BonusManager {
    
    companion object {
        private const val TAG = "BonusManager"
        private const val BONUS_API_BASE = "http://10.0.2.2:3007" // Android emulator localhost
        private const val MIN_PURCHASE_BONUS = 10000.0
    }
    
    /**
     * Check if purchase qualifies for bonus and create bonus record
     */
    suspend fun checkPurchaseBonus(
        walletAddress: String,
        userId: String,
        purchaseAmount: Double
    ): BonusResult {
        return withContext(Dispatchers.IO) {
            try {
                if (purchaseAmount < MIN_PURCHASE_BONUS) {
                    return@withContext BonusResult.NotQualified(
                        "Purchase amount too low. Minimum: $MIN_PURCHASE_BONUS USDTg"
                    )
                }
                
                val url = URL("$BONUS_API_BASE/api/bonus/create")
                val connection = url.openConnection() as HttpURLConnection
                
                connection.apply {
                    requestMethod = "POST"
                    setRequestProperty("Content-Type", "application/json")
                    doOutput = true
                }
                
                val requestBody = JSONObject().apply {
                    put("wallet_address", walletAddress)
                    put("user_id", userId)
                    put("purchase_amount", purchaseAmount)
                }
                
                OutputStreamWriter(connection.outputStream).use { writer ->
                    writer.write(requestBody.toString())
                    writer.flush()
                }
                
                val responseCode = connection.responseCode
                if (responseCode == HttpURLConnection.HTTP_OK) {
                    val response = BufferedReader(InputStreamReader(connection.inputStream))
                        .use { it.readText() }
                    
                    val json = JSONObject(response)
                    val success = json.getBoolean("success")
                    
                    if (success) {
                        val bonusId = json.getString("bonus_id")
                        val bonusAmount = json.getDouble("bonus_amount")
                        val tier = json.getString("tier")
                        val bonusPercent = json.getDouble("bonus_percent")
                        val message = json.getString("message")
                        
                        Log.d(TAG, "🎁 BONUS EARNED!")
                        Log.d(TAG, "💰 Amount: $bonusAmount USDTg")
                        Log.d(TAG, "🏆 Tier: $tier")
                        Log.d(TAG, "📝 $message")
                        
                        BonusResult.Success(
                            bonusId = bonusId,
                            bonusAmount = bonusAmount,
                            tier = tier,
                            bonusPercent = bonusPercent,
                            message = message
                        )
                    } else {
                        val error = json.optString("error", "Unknown error")
                        Log.e(TAG, "❌ Bonus failed: $error")
                        BonusResult.Error(error)
                    }
                } else {
                    Log.e(TAG, "❌ HTTP Error: $responseCode")
                    BonusResult.Error("HTTP Error: $responseCode")
                }
            } catch (e: Exception) {
                Log.e(TAG, "❌ Bonus check failed", e)
                BonusResult.Error(e.message ?: "Unknown error")
            }
        }
    }
    
    /**
     * Calculate bonus for purchase amount
     */
    suspend fun calculateBonus(purchaseAmount: Double): BonusCalculation? {
        return withContext(Dispatchers.IO) {
            try {
                val url = URL("$BONUS_API_BASE/api/bonus/calculate?amount=$purchaseAmount")
                val connection = url.openConnection() as HttpURLConnection
                
                connection.apply {
                    requestMethod = "GET"
                    setRequestProperty("Content-Type", "application/json")
                }
                
                val responseCode = connection.responseCode
                if (responseCode == HttpURLConnection.HTTP_OK) {
                    val response = BufferedReader(InputStreamReader(connection.inputStream))
                        .use { it.readText() }
                    
                    val json = JSONObject(response)
                    val success = json.getBoolean("success")
                    
                    if (success) {
                        BonusCalculation(
                            purchaseAmount = json.getDouble("purchase_amount"),
                            bonusAmount = json.getDouble("bonus_amount"),
                            tier = json.getString("tier"),
                            bonusPercent = json.getDouble("bonus_percent"),
                            qualifies = json.getBoolean("qualifies"),
                            minimumPurchase = json.getDouble("minimum_purchase")
                        )
                    } else {
                        null
                    }
                } else {
                    null
                }
            } catch (e: Exception) {
                Log.e(TAG, "❌ Bonus calculation failed", e)
                null
            }
        }
    }
    
    /**
     * Get user bonus statistics
     */
    suspend fun getUserStats(walletAddress: String): UserBonusStats? {
        return withContext(Dispatchers.IO) {
            try {
                val url = URL("$BONUS_API_BASE/api/bonus/user-stats?wallet_address=$walletAddress")
                val connection = url.openConnection() as HttpURLConnection
                
                connection.apply {
                    requestMethod = "GET"
                    setRequestProperty("Content-Type", "application/json")
                }
                
                val responseCode = connection.responseCode
                if (responseCode == HttpURLConnection.HTTP_OK) {
                    val response = BufferedReader(InputStreamReader(connection.inputStream))
                        .use { it.readText() }
                    
                    val json = JSONObject(response)
                    val success = json.getBoolean("success")
                    
                    if (success) {
                        UserBonusStats(
                            walletAddress = json.getString("wallet_address"),
                            totalPurchases = json.getDouble("total_purchases"),
                            totalBonuses = json.getDouble("total_bonuses"),
                            bonusCount = json.getInt("bonus_count"),
                            highestTier = json.getString("highest_tier"),
                            isVip = json.getBoolean("is_vip")
                        )
                    } else {
                        null
                    }
                } else {
                    null
                }
            } catch (e: Exception) {
                Log.e(TAG, "❌ Stats fetch failed", e)
                null
            }
        }
    }
    
    /**
     * Get system bonus statistics
     */
    suspend fun getSystemStats(): SystemBonusStats? {
        return withContext(Dispatchers.IO) {
            try {
                val url = URL("$BONUS_API_BASE/api/bonus/system-stats")
                val connection = url.openConnection() as HttpURLConnection
                
                connection.apply {
                    requestMethod = "GET"
                    setRequestProperty("Content-Type", "application/json")
                }
                
                val responseCode = connection.responseCode
                if (responseCode == HttpURLConnection.HTTP_OK) {
                    val response = BufferedReader(InputStreamReader(connection.inputStream))
                        .use { it.readText() }
                    
                    val json = JSONObject(response)
                    val success = json.getBoolean("success")
                    
                    if (success) {
                        SystemBonusStats(
                            totalBonuses = json.getLong("total_bonuses"),
                            totalBonusValue = json.getDouble("total_bonus_value"),
                            totalPurchaseVolume = json.getDouble("total_purchase_volume"),
                            bonusesToday = json.getInt("bonuses_today")
                        )
                    } else {
                        null
                    }
                } else {
                    null
                }
            } catch (e: Exception) {
                Log.e(TAG, "❌ System stats fetch failed", e)
                null
            }
        }
    }
}

// Result classes
sealed class BonusResult {
    data class Success(
        val bonusId: String,
        val bonusAmount: Double,
        val tier: String,
        val bonusPercent: Double,
        val message: String
    ) : BonusResult()
    
    data class NotQualified(val message: String) : BonusResult()
    data class Error(val message: String) : BonusResult()
}

data class BonusCalculation(
    val purchaseAmount: Double,
    val bonusAmount: Double,
    val tier: String,
    val bonusPercent: Double,
    val qualifies: Boolean,
    val minimumPurchase: Double
)

data class UserBonusStats(
    val walletAddress: String,
    val totalPurchases: Double,
    val totalBonuses: Double,
    val bonusCount: Int,
    val highestTier: String,
    val isVip: Boolean
)

data class SystemBonusStats(
    val totalBonuses: Long,
    val totalBonusValue: Double,
    val totalPurchaseVolume: Double,
    val bonusesToday: Int
)

File: wallet/bonus-manager.js
/*
==============================================
 File:        bonus-manager.js
 Author:      USDTG GROUP TECHNOLOGY LLC
 Created:     2025-10-13
 Last Update: 2025-10-13
 Version:     1.0

 Description:
   USDTgVerse Web Wallet Bonus Manager
   
   JavaScript client for bonus system:
   - 10,000 USDTg → +10 USDTg bonus
   - VIP tier system (Bronze to Diamond)
   - Automatic bonus notifications

 License:
   MIT License
==============================================
*/

class BonusManager {
    constructor() {
        this.apiBase = 'http://localhost:3007';
        this.minPurchaseBonus = 10000.0;
        this.tiers = {
            bronze: 10000,
            silver: 50000,
            gold: 100000,
            platinum: 500000,
            diamond: 1000000
        };
    }

    /**
     * Check if purchase qualifies for bonus and create bonus record
     */
    async checkPurchaseBonus(walletAddress, userId, purchaseAmount) {
        try {
            if (purchaseAmount < this.minPurchaseBonus) {
                console.log(`Purchase amount too low. Minimum: ${this.minPurchaseBonus} USDTg`);
                return {
                    qualified: false,
                    message: `Minimum purchase for bonus: ${this.minPurchaseBonus} USDTg`
                };
            }

            const response = await fetch(`${this.apiBase}/api/bonus/create`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    wallet_address: walletAddress,
                    user_id: userId,
                    purchase_amount: purchaseAmount
                })
            });

            const data = await response.json();

            if (data.success) {
                console.log('🎁 BONUS EARNED!');
                console.log('💰 Amount:', data.bonus_amount, 'USDTg');
                console.log('🏆 Tier:', data.tier);
                console.log('📝', data.message);

                this.showBonusNotification(
                    data.bonus_amount,
                    data.tier,
                    data.bonus_percent,
                    data.message
                );

                return {
                    qualified: true,
                    bonusId: data.bonus_id,
                    bonusAmount: data.bonus_amount,
                    tier: data.tier,
                    bonusPercent: data.bonus_percent,
                    message: data.message
                };
            } else {
                console.error('❌ Bonus failed:', data.error);
                return {
                    qualified: false,
                    error: data.error
                };
            }
        } catch (error) {
            console.error('❌ Bonus check failed:', error);
            return {
                qualified: false,
                error: error.message
            };
        }
    }

    /**
     * Calculate bonus for purchase amount
     */
    async calculateBonus(purchaseAmount) {
        try {
            const response = await fetch(
                `${this.apiBase}/api/bonus/calculate?amount=${purchaseAmount}`
            );

            const data = await response.json();

            if (data.success) {
                return {
                    success: true,
                    purchaseAmount: data.purchase_amount,
                    bonusAmount: data.bonus_amount,
                    tier: data.tier,
                    bonusPercent: data.bonus_percent,
                    qualifies: data.qualifies,
                    minimumPurchase: data.minimum_purchase,
                    tierThresholds: data.tier_thresholds
                };
            } else {
                return {
                    success: false,
                    error: data.error
                };
            }
        } catch (error) {
            console.error('❌ Bonus calculation failed:', error);
            return {
                success: false,
                error: error.message
            };
        }
    }

    /**
     * Get user bonus statistics
     */
    async getUserStats(walletAddress) {
        try {
            const response = await fetch(
                `${this.apiBase}/api/bonus/user-stats?wallet_address=${walletAddress}`
            );

            const data = await response.json();

            if (data.success) {
                return {
                    success: true,
                    walletAddress: data.wallet_address,
                    totalPurchases: data.total_purchases,
                    totalBonuses: data.total_bonuses,
                    bonusCount: data.bonus_count,
                    highestTier: data.highest_tier,
                    isVip: data.is_vip,
                    firstBonus: data.first_bonus,
                    lastBonus: data.last_bonus
                };
            } else {
                return {
                    success: false,
                    error: data.error
                };
            }
        } catch (error) {
            console.error('❌ Stats fetch failed:', error);
            return {
                success: false,
                error: error.message
            };
        }
    }

    /**
     * Get system bonus statistics
     */
    async getSystemStats() {
        try {
            const response = await fetch(`${this.apiBase}/api/bonus/system-stats`);
            const data = await response.json();

            if (data.success) {
                return {
                    success: true,
                    totalBonuses: data.total_bonuses,
                    totalBonusValue: data.total_bonus_value,
                    totalPurchaseVolume: data.total_purchase_volume,
                    bonusesToday: data.bonuses_today,
                    tierBreakdown: data.tier_breakdown
                };
            } else {
                return {
                    success: false,
                    error: data.error
                };
            }
        } catch (error) {
            console.error('❌ System stats fetch failed:', error);
            return {
                success: false,
                error: error.message
            };
        }
    }

    /**
     * Show bonus notification
     */
    showBonusNotification(bonusAmount, tier, bonusPercent, message) {
        const tierColors = {
            'Bronze': 'linear-gradient(135deg, #CD7F32 0%, #B87333 100%)',
            'Silver': 'linear-gradient(135deg, #C0C0C0 0%, #A8A8A8 100%)',
            'Gold': 'linear-gradient(135deg, #FFD700 0%, #FFA500 100%)',
            'Platinum': 'linear-gradient(135deg, #E5E4E2 0%, #B0C4DE 100%)',
            'Diamond': 'linear-gradient(135deg, #B9F2FF 0%, #00CED1 100%)'
        };

        const tierEmojis = {
            'Bronze': '🥉',
            'Silver': '🥈',
            'Gold': '🥇',
            'Platinum': '💎',
            'Diamond': '💠'
        };

        const notification = document.createElement('div');
        notification.className = 'bonus-notification';
        notification.innerHTML = `
            <div style="background: ${tierColors[tier] || tierColors['Bronze']}; 
                        color: white; padding: 25px; border-radius: 15px; 
                        box-shadow: 0 15px 50px rgba(0,0,0,0.4); 
                        position: fixed; top: 20px; right: 20px; 
                        max-width: 450px; z-index: 10000; animation: bonusSlideIn 0.6s cubic-bezier(0.68, -0.55, 0.265, 1.55);">
                <h3 style="margin: 0 0 15px 0; font-size: 24px; font-weight: bold;">
                    ${tierEmojis[tier]} ${tier} Tier Bonus!
                </h3>
                <p style="margin: 0 0 15px 0; font-size: 16px;">${message}</p>
                <div style="background: rgba(255,255,255,0.25); padding: 15px; border-radius: 10px; margin-top: 15px;">
                    <div style="display: flex; justify-content: space-between; margin-bottom: 10px;">
                        <span style="font-size: 14px;">💰 Bonus Amount:</span>
                        <span style="font-size: 18px; font-weight: bold;">${bonusAmount} USDTg</span>
                    </div>
                    <div style="display: flex; justify-content: space-between; margin-bottom: 10px;">
                        <span style="font-size: 14px;">🏆 Tier:</span>
                        <span style="font-size: 16px; font-weight: bold;">${tier}</span>
                    </div>
                    <div style="display: flex; justify-content: space-between;">
                        <span style="font-size: 14px;">📊 Bonus Rate:</span>
                        <span style="font-size: 16px; font-weight: bold;">${bonusPercent}%</span>
                    </div>
                </div>
                <div style="margin-top: 15px; padding: 10px; background: rgba(255,255,255,0.15); border-radius: 8px; text-align: center;">
                    <p style="margin: 0; font-size: 13px;">🎯 Keep purchasing to unlock higher tiers!</p>
                </div>
            </div>
        `;

        document.body.appendChild(notification);

        setTimeout(() => {
            notification.style.animation = 'bonusSlideOut 0.5s';
            setTimeout(() => notification.remove(), 500);
        }, 10000);
    }

    /**
     * Show tier progress indicator
     */
    showTierProgress(currentPurchase, currentTier) {
        const nextTierThreshold = this.getNextTierThreshold(currentTier);
        if (!nextTierThreshold) return;

        const progress = (currentPurchase / nextTierThreshold) * 100;
        const remaining = nextTierThreshold - currentPurchase;

        console.log(`📊 Tier Progress: ${progress.toFixed(2)}%`);
        console.log(`🎯 Next Tier: ${remaining.toFixed(2)} USDTg remaining`);
    }

    /**
     * Get next tier threshold
     */
    getNextTierThreshold(currentTier) {
        const tierOrder = ['Bronze', 'Silver', 'Gold', 'Platinum', 'Diamond'];
        const currentIndex = tierOrder.indexOf(currentTier);
        
        if (currentIndex === -1 || currentIndex === tierOrder.length - 1) {
            return null;
        }

        const nextTier = tierOrder[currentIndex + 1];
        return this.tiers[nextTier.toLowerCase()];
    }
}

// Add CSS animations
const style = document.createElement('style');
style.textContent = `
    @keyframes bonusSlideIn {
        from {
            transform: translateX(500px) scale(0.8);
            opacity: 0;
        }
        to {
            transform: translateX(0) scale(1);
            opacity: 1;
        }
    }
    
    @keyframes bonusSlideOut {
        from {
            transform: translateX(0) scale(1);
            opacity: 1;
        }
        to {
            transform: translateX(500px) scale(0.8);
            opacity: 0;
        }
    }
`;
document.head.appendChild(style);

// Export for use
if (typeof module !== 'undefined' && module.exports) {
    module.exports = BonusManager;
}

